### PR TITLE
[3.5] Fix issue, when we save the same customers at the same time.

### PR DESCRIPTION
**Pull request by van-vanich**
## Pull Request description

issue: https://github.com/thingsboard/thingsboard/issues/6291
PE: https://github.com/thingsboard/thingsboard-pe/pull/795

When we save the same customers at the same time, SQL doesn't check it. And when we use function findCustomerByTenantIdAndTitle() we have error:
org.springframework.dao.IncorrectResultSizeDataAccessException: query did not return a unique result: 2; nested exception is javax.persistence.NonUniqueResultException: query did not return a unique result: 2

We fix this issue and add a script for the upgrade of old name customers.
Order customer name by created time. 
Example upgrade customer name with the same name:
Before:
NAME(created time = 1), NAME(created time = 2), NAME(created time = 3)
After:
NAME(created time = 1), NAME-1(created time = 2), NAME-2 (created time = 3)

## General checklist

- [x] This PR has the corresponding CE PR or contains unique code changes related to PE only (does not require same changes in CE).
- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

**Comment by ashvayka** [comment]
We can't merge this at the moment, since there is no backward compatibility. Someone may already have two customers with the same title.

**Comment by ashvayka** [comment]
I suggest adding a script that automatically renames the conflicting titles. We will consider merging this to 3.4.

File: application/src/main/java/org/thingsboard/server/install/ThingsboardInstallService.java [hunk 1]
```diff
@@ -220,7 +220,11 @@ public void performInstall() {
                             log.info("Updating system data...");
                             systemDataLoaderService.updateSystemWidgets();
                             break;
```
Comment (dmytro-landiak) [inline application/src/main/java/org/thingsboard/server/install/ThingsboardInstallService.java:222]: remove this "break"

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 2]
```diff
@@ -547,4 +555,63 @@ private void updateOAuth2Params() {
...
+                }
+
+                @Override
+                protected void updateEntity(Tenant entity) {
+                    updateDuplicateCustomersTitle(entity);
+                }
+            };
+
+    private List<Customer> updateDuplicateCustomersTitle(Tenant entity) {
+        customerDao.find(entity.getId());
```
Comment (dmytro-landiak) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:578]: duplicate

File: dao/src/test/java/org/thingsboard/server/dao/service/BaseCustomerServiceTest.java [hunk 3]
```diff
@@ -245,4 +253,49 @@ public void testFindCustomersByTenantIdAndTitle() {
...
+                    }
+                }
+                finishLatch.countDown();
+            });
+        }
+        assertTrue(readyLatch.await(TIMEOUT, TimeUnit.SECONDS));
+        Thread.yield();
+        startLatch.countDown(); //run all-at-once submitted tasks
+        assertTrue(finishLatch.await(TIMEOUT, TimeUnit.SECONDS));
+        Assert.assertEquals("Saved count not equals  1", cntSaved.get(), 1);
```
Comment (dmytro-landiak) [inline dao/src/test/java/org/thingsboard/server/dao/service/BaseCustomerServiceTest.java:296]: wrong message. should probably be "Save count equals 1"

File: dao/src/test/java/org/thingsboard/server/dao/service/BaseCustomerServiceTest.java [hunk 4]
```diff
@@ -245,4 +253,49 @@ public void testFindCustomersByTenantIdAndTitle() {
         Assert.assertFalse(pageData.hasNext());
         Assert.assertEquals(0, pageData.getData().size());
     }
+
+    @Test
+    public void testCreateTwoCustomerInSameTime() throws InterruptedException {
+        final int TIMEOUT = 10;
+        int parallelCount = 3;
+        ExecutorService executorService = Executors.newFixedThreadPool(parallelCount, ThingsBoardThreadFactory.forName(getClass().getSimpleName() + "-test-scope"));
```
Comment (dmytro-landiak) [inline dao/src/test/java/org/thingsboard/server/dao/service/BaseCustomerServiceTest.java:261]: shutdown after the test in finally block for example

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 5]
```diff
@@ -547,4 +555,63 @@ private void updateOAuth2Params() {
...
+    }
+
+    protected Customer updateCustomerTitle(TenantId id, Customer customer) {
+        customerDao.removeById(id, customer.getUuidId());
+        return customerDao.save(id, customer);
+    }
+
+    protected void sortCustomersByTenantIdAndTitleAndCreatedTime(List<Customer> customers) {
+        customers.sort((o1, o2) -> {
+            if (!o1.getTenantId().equals(o2.getTenantId())) return o1.getTenantId().toString().compareTo(o2.getTenantId().toString());
```
Comment (dmytro-landiak) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:611]: do you need to do this if all customers are related to single tenant here?

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 6]
```diff
@@ -547,4 +555,63 @@ private void updateOAuth2Params() {
...
+
+    protected List<Customer> updateDuplicateCustomersTitle(List<Customer> customers) {
+        if (customers == null || customers.isEmpty()) return customers;
+        sortCustomersByTenantIdAndTitleAndCreatedTime(customers);
+        int countEqualsTitleAndTenantIdBefore = 0;
+        String lastCustomerName = customers.get(0).getName();
+        TenantId lastTenantId = customers.get(0).getTenantId();
+        for (int i = 1; i < customers.size(); i++) {
+            Customer customer = customers.get(i);
+            if (customer.getName().equals(lastCustomerName) && customer.getTenantId().equals(lastTenantId)) {
```
Comment (dmytro-landiak) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:591]: && customer.getTenantId().equals(lastTenantId) not needed ?

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 7]
```diff
@@ -547,4 +555,63 @@ private void updateOAuth2Params() {
...
+            } else {
+                countEqualsTitleAndTenantIdBefore = 0;
+            }
+            lastCustomerName = customer.getName();
+            lastTenantId = customer.getTenantId();
+        }
+        return customers;
+    }
+
+    protected Customer updateCustomerTitle(TenantId id, Customer customer) {
```
Comment (dmytro-landiak) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:604]: why remove? simple save not enough?

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java (hunk 5 above)
Comment (van-vanich) [reply application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:611]: Yes, I'm testing, and this code returns entities with a different tenantId

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java (hunk 6 above)
Comment (van-vanich) [reply application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:591]: The list contains customers with different tenantId

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java (hunk 5 above)
Comment (van-vanich) [reply application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:611]: Sorry, my mistake. Thanks for advice!

**Comment by dmytro-landiak** [comment]
LGTM!

**Review by dmytro-landiak — APPROVED** [review]

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 8]
```diff
@@ -547,4 +557,69 @@ private void updateOAuth2Params() {
         log.warn("CAUTION: Update of Oauth2 parameters from 3.2.2 to 3.3.0 available only in ThingsBoard versions 3.3.0/3.3.1");
     }
 
+    private final PaginatedUpdater<String, Tenant> tenantCustomersTitleUpdater =
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:560]: Create a separate component for this updater and cover it with unit tests.

**Comment by ashvayka** [comment]
@van-vanich please fix the conflicts and improve the upgrade script. The existing upgrade script will work for minutes on our Demo/PaaS environments. We need to find all duplicated customers with one DB call instead of iterating through all the tenants and performing one DB call per tenant.

**Comment by ashvayka** [comment]
Additionally, separate PR for PE is required.

File: dao/src/main/java/org/thingsboard/server/dao/customer/CustomerServiceImpl.java [hunk 9]
```diff
@@ -98,9 +101,18 @@ public ListenableFuture<Customer> findCustomerByIdAsync(TenantId tenantId, Custo
...
         customerValidator.validate(customer, Customer::getTenantId);
-        Customer savedCustomer = customerDao.save(customer.getTenantId(), customer);
-        dashboardService.updateCustomerDashboards(savedCustomer.getTenantId(), savedCustomer.getId());
-        return savedCustomer;
+        try {
+            Customer savedCustomer = customerDao.save(customer.getTenantId(), customer);
+            dashboardService.updateCustomerDashboards(savedCustomer.getTenantId(), savedCustomer.getId());
+            return savedCustomer;
+        } catch (Exception t) {
+            ConstraintViolationException e = extractConstraintViolationException(t).orElse(null);
```
Comment (ashvayka) [inline dao/src/main/java/org/thingsboard/server/dao/customer/CustomerServiceImpl.java:109]: if you add this, then remove the redundant check from the customer validator.

File: dao/src/main/java/org/thingsboard/server/dao/sql/customer/CustomerRepository.java [hunk 10]
```diff
@@ -43,4 +43,10 @@ Page<CustomerEntity> findByTenantId(@Param("tenantId") UUID tenantId,
     @Query("SELECT externalId FROM CustomerEntity WHERE id = :id")
     UUID getExternalIdById(@Param("id") UUID id);
 
+
+    @Query(value = "SELECT c.* FROM customer c " +
+            "INNER JOIN (SELECT tenant_id, title, FROM customer GROUP BY tenant_id, title HAVING COUNT(title) > 1) dc " +
+            "ON c.tenant_id = :tenantId AND c.tenant_id = dc.tenant_id AND c.title = dc.title",
+            nativeQuery = true)
+    Page<CustomerEntity> findCustomerWithEqualTitle(@Param("tenantId") UUID tenantId, Pageable pageable);
```
Comment (ashvayka) [inline dao/src/main/java/org/thingsboard/server/dao/sql/customer/CustomerRepository.java:51]: No need to filter by tenant id. It is better to sort by tenant id.

File: application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java [hunk 11]
```diff
@@ -564,6 +572,24 @@ private void updateOAuth2Params() {
...
+            new PaginatedUpdater<>() {
+
+                @Override
+                protected String getName() {
+                    return "Customer title updater";
+                }
+
+                @Override
+                protected PageData<Tenant> findEntities(String id, PageLink pageLink) {
+                    return tenantService.findTenants(pageLink);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:585]: This is wrong. Why do you need to iterate over tenants? Just fetch ALL customers with duplicated names (grouped by tenant) and iterate over this list. If we have 50K tenants in the system, we will need to run 50K separate queries (1 per tenant). The goal is to run only one query.

Comment (van-vanich) [reply application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java:585]: Thanks for review PR, 
I add commit with improve for this case

File: application/src/main/java/org/thingsboard/server/service/install/update/TenantCustomersTitleUpdaterComponent.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    protected List<Customer> updateDuplicateCustomersTitle(List<Customer> customers) {
+        if (customers == null || customers.isEmpty()) return customers;
+        sortCustomersByTitleAndCreatedTime(customers);
+        int countEqualsTitleAndTenantIdBefore = 0;
+        String lastCustomerName = customers.get(0).getName();
+        for (int i = 1; i < customers.size(); i++) {
+            Customer customer = customers.get(i);
+            if (customer.getName().equals(lastCustomerName)) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/install/update/TenantCustomersTitleUpdaterComponent.java:61]: make sure that the tenantId is not changed. If two tenants have the same customer names - it is ok.

**Pull request closed without merging**